piet-gpu-hal/examples/shader/stack.comp

```diff
@@ -0,0 +1,217 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Implementation of stack monoid (companion to blog post)
+
+#version 450
+
+#define LG_PART_SIZE 4
+#define PART_SIZE (1 << LG_PART_SIZE)
+
+layout(local_size_x = PART_SIZE) in;
+
+layout(set = 0, binding = 0) buffer DataBuf {
+    uint[] data;
+};
+
+layout(set = 0, binding = 1) volatile buffer StateBuf {
+    uint[] state;
+};
+
+#define STATE_COUNTER 0
+#define STATE_FLAGS 1
+#define STATE_NPOP_LOCAL 2
+#define STATE_NPUSH_LOCAL 3
+#define STATE_NPUSH_PREFIX 4
+#define STATE_SIZE 6
+
+#define FLAG_INIT 0
+#define FLAG_LOCAL 1
+#define FLAG_DONE 2
+
+layout(set = 0, binding = 2) volatile buffer StackBuf {
+    uint[] scratch;
+};
+
+shared uint sh_part_ix;
+shared uint sh_flags;
+shared uint sh_npop[PART_SIZE];
+shared uint sh_npush[PART_SIZE];
+shared uint sh_stack[PART_SIZE];
+
+void main() {
+    uint th_ix = gl_LocalInvocationID.x;
+    // Determine partition by atomic counter.
+    if (th_ix == 0) {
+        sh_part_ix = atomicAdd(state[STATE_COUNTER], 1);
+    }
+    barrier();
+    uint part_ix = sh_part_ix;
+    uint el_ix = part_ix * PART_SIZE + th_ix;
+    uint inp = data[el_ix];
+
+    uint local_pop = 1 - inp;
+    uint npop = local_pop;
+    uint npush = inp;
+    uint stack = el_ix;
+    sh_npop[th_ix] = npop;
+    sh_npush[th_ix] = npush;
+    sh_stack[th_ix] = stack;
+    bool pending = true;
+
+    // Reduction tree for computation of monoid for this partition
+    for (uint j = 0; j < LG_PART_SIZE; j++) {
+        uint bit = 1 << j;
+        uint first = th_ix & (-2u * bit);
+        uint mid = first + bit - 1;
+        uint last = mid + bit;
+        // Combine monoid from first..first + bit and first + bit..first + bit * 2
+        // to first..first + bit * 2
+        barrier();
+        // In this section, only read from shared memory
+        uint push_mid = sh_npush[mid];
+        if ((th_ix & bit) != 0) {
+            uint m = max(push_mid, npop);
+            npush = m - npop + npush;
+            uint prepop = npop - local_pop;
+            npop = sh_npop[mid] + m - push_mid;
+            if (pending && push_mid > prepop) {
+                // Generate final output when it can be resolved in this partition
+                data[el_ix] = sh_stack[mid - prepop];
+                pending = false;
+            }
+        }
+        uint pop_last = sh_npop[last];
+        uint m = max(push_mid, pop_last);
+        uint push_last = sh_npush[last];
+        if (last - th_ix >= push_last) {
+            uint iprev = last - th_ix - push_last;
+            if (iprev < m - pop_last) {
+                stack = sh_stack[mid - pop_last - iprev];
+            } else {
+                // TODO: not needed
+                stack = 99;
```

Comment:
If you plan to leave this, a comment saying how it can be removed would be nice.